### PR TITLE
core: thread: allocate RPC arg data on demand

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Delays the RPC arg data allocation until it's needed to avoid doing RPC
when the caller in normal world isn't prepared to handle it.

Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU)
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

Should fix https://github.com/OP-TEE/optee_os/issues/1401

### CODE REVIEW

**Comment by jenswi-linaro** [comment]
Any thoughts about this? Should it be done differently?

File: core/arch/arm/kernel/thread.c [hunk 1]
```diff
@@ -590,34 +590,18 @@ void thread_handle_std_smc(struct thread_smc_args *args)
...
-	}
 
 	thread_std_smc_handler_ptr(args);
 
-	tee_fs_rpc_cache_clear(&thr->tsd);
-	if (!thread_prealloc_rpc_cache) {
-		thread_rpc_free_arg(thr->rpc_carg);
-		thr->rpc_carg = 0;
-		thr->rpc_arg = 0;
+	if (args->a0 == OPTEE_SMC_RETURN_OK) {
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread.c]: why is rpc cache cleaned only on successful SMCs ?

Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: A successful SMC is a recognized SMC, so that means that RPC should be OK.

**Comment by jenswi-linaro** [comment]
Should we drop this or not?

File: core/arch/arm/kernel/thread.c [hunk 2]
```diff
@@ -1285,24 +1212,102 @@ void thread_rpc_alloc_arg(size_t size, paddr_t *arg, uint64_t *cookie)
...
+
+	if (!arg) {
+		paddr_t p;
+		uint64_t c;
+
+		thread_rpc_alloc_arg(
+			OPTEE_MSG_GET_ARG_SIZE(THREAD_RPC_MAX_NUM_PARAMS),
+			&p, &c);
+		if (!p || !ALIGNMENT_IS_OK(p, struct optee_msg_arg))
+			return false;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread.c]: Missing free of the allocated memory when it does not fulfil the alignment constraint.

**Comment by jenswi-linaro** [comment]
Update

File: core/arch/arm/kernel/thread.c [hunk 3]
```diff
@@ -1285,24 +1212,105 @@ void thread_rpc_alloc_arg(size_t size, paddr_t *arg, uint64_t *cookie)
...
+			struct optee_msg_arg **arg_ret, uint64_t *carg_ret,
+			struct optee_msg_param **params_ret)
+{
+	struct thread_ctx *thr = threads + thread_get_id();
+	struct optee_msg_arg *arg = thr->rpc_arg;
+	size_t sz = OPTEE_MSG_GET_ARG_SIZE(THREAD_RPC_MAX_NUM_PARAMS);
+	paddr_t p;
+	uint64_t c;
+
+
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread.c]: I'll fix.

**Comment by jenswi-linaro** [comment]
Comment addressed, tags applied and rebased.

**Comment by dryvenn** [comment]
Sorry guys I'm a few minutes late, but for what it's worth, the fix indeed solves #1401. Thanks @jenswi-linaro!

**Comment by jforissier** [comment]
@dryvenn no problem, good to know it's working for you!

**Pull request merged**